crates/compose/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+/// composer will use an export of import name to satisfy the import. The
+/// composer will then alias the export of the dependency to the import of the
+/// dependent component. Finally, the composer will export all exports from the
+/// dependent component to its dependents. The composer will then encode the
+/// composition graph into a byte array and return it.
+pub async fn compose<'a, L: ComponentLoader>(
+    loader: &'a L,
+    component: &LockedComponent,
+) -> anyhow::Result<Vec<u8>> {
+    Composer::new(loader).compose(component).await
+}
+
+/// This trait is used to load component source code from a locked component source across various embdeddings.
+#[async_trait::async_trait]
+pub trait ComponentLoader {
+    async fn load_component(
+        &self,
+        source: &locked::LockedComponentSource,
+    ) -> anyhow::Result<Vec<u8>>;
+}
+
+struct Composer<'a, L> {
+    graph: CompositionGraph,
+    loader: &'a L,
+}
+
+impl<'a, L: ComponentLoader> Composer<'a, L> {
+    async fn compose(mut self, component: &LockedComponent) -> anyhow::Result<Vec<u8>> {
+        let source = self.loader.load_component(&component.source).await?;
+
+        if component.dependencies.is_empty() {
+            return Ok(source);
+        }
+
+        let (world_id, instantiation_id) = self.register_package(&component.id, None, source)?;
+
+        let prepared = self.prepare_dependencies(world_id, component).await?;
+
+        let arguments = self
+            .build_instantiation_arguments(world_id, prepared)
+            .await?;
+
+        for (argument_name, argument) in arguments {
+            self.graph
+                .set_instantiation_argument(instantiation_id, &argument_name, argument)?;
+        }
+
+        self.export_dependents_exports(world_id, instantiation_id)?;
+
+        self.graph
+            .encode(Default::default())
+            .with_context(|| "failed to encode composition graph")
+            .map(Into::into)
+    }
+
+    fn new(loader: &'a L) -> Self {
+        Self {
+            graph: CompositionGraph::new(),
+            loader,
+        }
+    }
+
+    async fn prepare_dependencies(
+        &mut self,
+        world_id: WorldId,
+        component: &LockedComponent,
+    ) -> anyhow::Result<IndexMap<String, DependencyInfo>> {
+        let mut imports = self.graph.types()[world_id].imports.clone();
+
+        let mut prepared = IndexMap::new();
+
+        // TODO: collect and return conflicts
+
+        for (dependency_name, dependency) in component.dependencies.iter() {
+            // As we match dependency names against import names, we remove the
+            // matched import name from the imports map to prevent duplicate
+            // matches. Collecting the import keys here is an optimization to
+            // avoid walking the full set of import keys on each iteration.
+            let import_keys = imports.keys().cloned().collect::<Vec<_>>();
+
+            let mut matched = Vec::new();
+
```

Comment:
I think we also want to keep track of the imports of the dependencies. The imports themselves should not have any dependencies. Right now, we just float those dependencies through and you end up with a linker error.

Reply:
I don't think it's quite that easy. There are some instances (in-fact likely the common case) where we want to float dependencies of dependencies through. This is where a target worlds check would be applied subsequently to composition to ensure that composition conforms to what Spin knows how to handle.

Reply:
Sounds good - I think we'll want to eventually work on improving how linker errors get bubbled to users anyway, and if we improve that, this concern goes away.